File: src/main/java/org/pl/lexer/tokenizer/NameTokenizer.java
package org.pl.lexer.tokenizer;

import org.pl.lexer.ILexer;
import org.pl.lexer.exception.TokenizerException;
import org.pl.lexer.token.IToken;
import org.pl.lexer.token.NameToken;
import org.pl.lexer.token.keyword.*;


/**
 * Tokenizes names and reserved keywords. If no keyword matches, a name token
 * is returned.
 */
public class NameTokenizer implements ITokenizer {

    @Override
    public boolean matches(Character c) {
        return Character.isAlphabetic(c);
    }

    @Override
    public IToken tokenize(ILexer lexer) throws TokenizerException {
        var nameBuilder = new StringBuilder();

        while (!lexer.isEndOfProgram() && matches(lexer.getSymbol())) {
            nameBuilder.append(lexer.getSymbol());
            lexer.advanceCursor();
        }
        lexer.advanceCursor(-1);

        var name = nameBuilder.toString();
        switch (name) {
            case "var":
                return new VarKeywordToken();
            case "const":
                return new ConstKeywordToken();
            case "if":
                return new IfKeywordToken();
            case "else":
                return new ElseKeywordToken();
            case "true":
                return new TrueKeywordToken();
            case "false":
                return new FalseKeywordToken();
            case "fun":
                return new FunctionKeywordToken();
            case "loop":
                return new LoopKeywordToken();
            case "break":
                return new BreakKeywordToken();
            case "continue":
                return new ContinueKeywordToken();
        }
        return new NameToken(name);
    }
}

File: src/main/java/org/pl/lexer/tokenizer/NameTokenizer.kt
package org.pl.lexer.tokenizer

import org.pl.lexer.ILexer
import org.pl.lexer.token.BoolToken
import org.pl.lexer.token.IToken
import org.pl.lexer.token.NameToken
import org.pl.lexer.token.NativeToken
import org.pl.lexer.token.keyword.*

/**
 * Tokenizes names and reserved keywords. If no keyword matches, a name token
 * is returned.
 */
class NameTokenizer : ITokenizer {

    override fun matches(c: Char) = c.isLetter()

    override fun tokenize(lexer: ILexer): IToken {
        val nameBuilder = StringBuilder()
        while (!lexer.isEndOfProgram && matches(lexer.symbol)) {
            nameBuilder.append(lexer.symbol)
            lexer.advanceCursor()
        }
        lexer.advanceCursor(-1)
        val name = nameBuilder.toString()
        when (name) {
            "var" -> return VarKeywordToken()
            "const" -> return ConstKeywordToken()
            "if" -> return ConditionalKeywordToken()
            "else" -> return ElseKeywordToken()
            "true" -> return BoolToken(true)
            "false" -> return BoolToken(false)
            "fun" -> return FunctionKeywordToken()
            "loop" -> return LoopKeywordToken()
            "break" -> return BreakKeywordToken()
            "continue" -> return ContinueKeywordToken()
            "number" -> return NumberKeywordToken()
            "bool" -> return BoolKeywordToken()
            "native" -> {
                // The native keyword needs some special treatment since the lexer has
                // to completely ignore this part
                val programCodeBuilder = StringBuilder()
                var braceCount = 1
                while (lexer.symbol != '{') {
                    lexer.advanceCursor()
                }
                while (true) {
                    lexer.advanceCursor()
                    if (lexer.symbol == '{') {
                        braceCount++
                    } else if (lexer.symbol == '}') {
                        braceCount--
                        if (braceCount <= 0) {
                            break
                        }
                    }
                    programCodeBuilder.append(lexer.symbol)
                }
                val programCode = programCodeBuilder.toString().trim()
                return NativeToken(programCode)
            }
        }
        return NameToken(name)
    }
}